### PR TITLE
feat(dragAndDrop): #MAG-392 implement drag and drop

### PR DESCRIPTION
**Pull request by alicedraillard**
## Describe your changes
user can d&d boards into either side bar or folder list
folders are hightlighted in grey (border color) when dragging over
right check is present
modal if no rights or confirmation needed

## Checklist tests
test rights depending on folder and board rights

## Issue ticket number and link
[MAG-392](https://jira.support-ent.fr/browse/MAG-392)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: frontend/src/components/folder-item/FolderItem.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  const [{ isOver }, drop] = useDrop({
+    accept: "board",
+    drop: () => setHasDrop(true),
+    collect: (monitor: any) => ({
+      isOver: !!monitor.isOver(),
+    }),
+  });
+
+  useEffect(() => {
+    const boards = dragAndDropBoards;
```
Comment (ThomasSevagen) [inline frontend/src/components/folder-item/FolderItem.tsx:56]: on peut peut être sortir le plus gros de la logique du useEffect :)

File: frontend/src/components/side-bar/SideBar.tsx [hunk 2]
```diff
@@ -69,6 +80,27 @@ export const SideBar: React.FunctionComponent<SideBarProps> = ({
     ).buildFolders(deletedFolders);
   }
 
+  const reducer = (
+    state: { fileList: string | any[] },
```
Comment (ThomasSevagen) [inline frontend/src/components/side-bar/SideBar.tsx:84]: problème de typage?

File: frontend/src/components/tree-view/TreeViewContainer.tsx [hunk 3]
```diff
@@ -1,32 +1,278 @@
...
+  };
+  const handleDragLeave = (e: React.ChangeEvent<HTMLInputElement>) => {
+    let targetElement: HTMLElement;
+    if (e.target.closest("li")?.id) {
+      if (
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.MY_BOARDS ||
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.PUBLIC_BOARDS ||
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.DELETED_BOARDS
+      ) {
+        console.log(e.target.closest("li")?.closest(".action-container"));
```
Comment (ThomasSevagen) [inline frontend/src/components/tree-view/TreeViewContainer.tsx:67]: log

File: frontend/src/components/tree-view/TreeViewContainer.tsx [hunk 4]
```diff
@@ -1,32 +1,278 @@
...
+
+  const handleDragOver = (e: React.ChangeEvent<HTMLInputElement>) => {
+    let targetElement: HTMLElement;
+    if (e.target.closest("li")?.id) {
+      if (
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.MY_BOARDS ||
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.PUBLIC_BOARDS ||
+        e.target.closest("li")?.id.split("_")[1] == FOLDER_TYPE.DELETED_BOARDS
+      ) {
+        console.log(e.target.closest("li")?.closest(".action-container"));
```
Comment (ThomasSevagen) [inline frontend/src/components/tree-view/TreeViewContainer.tsx:91]: log

File: frontend/src/components/tree-view/TreeViewContainer.tsx [hunk 5]
```diff
@@ -1,32 +1,278 @@
...
+      } as IFolderResponse);
+    } else {
+      clickedFolder =
+        folders.find((folder: Folder) => folder.id == folderId) ?? new Folder();
+    }
+
+    return clickedFolder;
+  };
+
+  const proceedOnDragAndDrop = async (
```
Comment (ThomasSevagen) [inline frontend/src/components/tree-view/TreeViewContainer.tsx:205]: cette fc a pas été définie quelque part déjà?

File: frontend/src/components/board-item/BoardItem.tsx [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        cursor: "move",
+      }}
+    >
+      <Card
+        app={currentApp!}
+        options={{
+          type: "board",
+          title: "",
+        }}
+        // onClick={() => {setIsToasterOpen()}}
```
Comment (FlorentMr) [inline frontend/src/components/board-item/BoardItem.tsx:80]: commentaire

File: frontend/src/components/folder-item/FolderItem.tsx [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    const dragAndDropBoardsIds: string[] = dragAndDropBoards.map(
+      (board: Board) => board._id,
+    );
+
+    resetDragAndDrop();
+    dragAndDropBoardsCall(dragAndDropBoardsIds, dragAndDropTargetId);
+  };
+
+  const handleNoRightsDragAndDrop = (): void => {
+    onSetModaleData({
```
Comment (FlorentMr) [inline frontend/src/components/folder-item/FolderItem.tsx:140]: Un peu franglais le "Modale", switch pour "Modal" = 👍👍

File: frontend/src/core/constants/rights.const.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+export const rights = {
```
Comment (FlorentMr) [inline frontend/src/core/constants/rights.const.ts:1]: On a déjà un fichier similaire : frontend/src/config/index.ts

**Review by ThomasSevagen — APPROVED** [review]

**Review by FlorentMr — APPROVED** [review]

**Pull request merged**